[JENKINS-42645] Case insensitive search by default for new and anonymous users

szpak opened the pull request:
# Description

See [JENKINS-42645](https://issues.jenkins-ci.org/browse/JENKINS-42645).

**In short**: Search is case insensitive by default (for newly created and anonymous users)

TODO: This PR changes behavior for newly created and anonymous users. It would be good to have it also for already existing users. Nevertheless it is more complicated and should be rather moved to another issue. Implementation suggestions are welcome: https://groups.google.com/d/msg/jenkinsci-dev/I9PVV3j0Eb8/Ln4njJlGDgAJ (or in [JENKINS-42701](https://issues.jenkins-ci.org/browse/JENKINS-42701)).

### Changelog entries

Proposed changelog entries:

* Entry 1: JENKINS-42645, Search is case insensitive by default (for newly created and anonymous users)

### Submitter checklist

- [x] JIRA issue is well described
- [x] Link to JIRA ticket in description, if appropriate
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc)

### Desired reviewers

@daniel-beck - in relation to #2683

szpak committed:
[JENKINS-42645] Case insensitive search by default

oleg-nenashev commented:
@jenkinsci/code-reviewers

szpak commented on core/src/main/java/hudson/search/FixedSet.java:
```diff
@@ -51,7 +51,7 @@ public void find(String token, List<SearchItem> result) {
                 name=name.toLowerCase();
             }
-            if(token.equals(i.getSearchName()))
+            if(token.equals(name))
```
By the fixed a bug which has been detected by tests after my change

oleg-nenashev replied:
Maybe deserves a separate JIRA ticket if it is important enough

szpak replied:
It's probably quite rare case, but I created [JENKINS-42960](https://issues.jenkins-ci.org/browse/JENKINS-42960)

daniel-beck replied:
This should use `equalsIgnoreCase` to not misbehave in Turkish locale.

szpak replied:
I see that under the hood `equalsIgnoreCase` uses `regionMatches` is seems to be smarter that just `toLowerCase()` without locale definition. In that case I can remove `toLowerCases` statements and use `equals` and `equalsIgnoreCase` cnditionally.

daniel-beck replied:
@szpak Yes, please do that.

szpak replied:
Changed (also a sibling `contains()` call) .

oleg-nenashev commented on core/src/main/java/hudson/search/UserSearchProperty.java:
```diff
@@ -11,7 +11,9 @@
 public class UserSearchProperty extends hudson.model.UserProperty {
-    
+
+    private static final boolean CASE_INSENSITIVE_SEARCH_MODE = true;
```
Maybe `DEFAULT_SEARCH_ID_CASE_INSENSITIVE`? If you add such static field, maybe it is useful to initialize the default value using `SystemProperties.getBoolean()`

szpak replied:
What does "ID" in your proposal stand? I don't fully get that name.

oleg-nenashev commented on core/src/test/java/hudson/model/ViewTest.java:
```diff
@@ -98,7 +99,7 @@ public void getAllItems() throws Exception {
 
         Mockito.when(leftView.getItems()).thenReturn(Arrays.asList(leftJob, sharedJob));
-        Mockito.when(rightView.getItems()).thenReturn(Arrays.asList(rightJob));
+        Mockito.when(rightView.getItems()).thenReturn(Collections.singletonList(rightJob));
```
unrelated, but OK

oleg-nenashev commented on core/src/test/java/jenkins/widgets/HistoryPageFilterTest.java:
```diff
@@ -326,40 +326,35 @@ public void test_search_runs_by_build_number() throws IOException {
-        //and
-        historyPageFilter.setSearchString("failure");
-        //and
+    public void should_be_case_insensitive_by_default() throws IOException {
```
Please add `@Issue("JENKINS-42645")`

szpak replied:
Done

oleg-nenashev reviewed (changes requested):
Looks good to me excepting the static variable name and missing support of properties. Since the default behavior changes, it may be useful for users to have an option to enable the legacy behavior by setting a system property

daniel-beck commented:
> it may be useful for users to have an option to enable the legacy behavior by setting a system property

@oleg-nenashev Do you actually expect a significant share of users to want case sensitive search by default? This seems like a "let's change it and consider adding an option if there's actual demand" kind of situation to me.

oleg-nenashev commented:
> Do you actually expect a significant share of users to want case sensitive search by default? This seems like a "let's change it and consider adding an option if there's actual demand" kind of situation to me.

I doubt that there will be a significant share, but de-facto we introduce a change without providing an option to restore the original one. And I have to confirm that on one of our instances there was "instance" case where letter cases did really matter (FOOBAR ad FooBar were valid and **different** users in the system).

Since it is an easy fix, I would rather propose to do it. In order to speedup the integration,  I can just create a new PR on the top of this one within a couple of hours @daniel-beck WDYT?

daniel-beck commented:
@oleg-nenashev 

> we introduce a change without providing an option to restore the original one

Yes, that's how changes work. It's rare that a change just switches to the other of two possible options, but we change things all the time.

Hidden options are only relevant when there's an experimental feature (something we've basically moved away from), an expectation that not having them will break things, or we introduce them in response to some user demand that still doesn't warrant a UI option. None of this is the case here. 

> one of our instances there was "instance" case where letter cases did really matter

That's why it can be configured on a per-user basis.

I don't think the additional complexity is warranted, and that implementing that would be a mistake -- but I won't reject it if you see it differently.

szpak commented:
In my opinion it's much more convenient to have case insensitive search (failure vs FAILURE) and particular (authorized) user can change it on demand. Nevertheless, it's not a big deal to just read an extra system property - just in case, so why not. @oleg-nenashev How do you propose to name that property? In code I see different naming conventions.

imod commented:
I think thats a very strange discussion - I now no tool where the default search behaviour is case sensitive...

daniel-beck reviewed (commented):
AFAIU the case insensitive mode will misbehave in Turkish locale – not introduced here, but touched on anyway, so should be changed.

Otherwise :+1: for the overall change. The previous default was undesirable in the vast majority of cases.

szpak committed:
[JENKINS-42960] Search in FixedSet more locale friendly

String.equalsIgnoreCase is safer than toLowerCase when non English
locales are used.

szpak committed:
[JENKINS-42645] Review remarks

daniel-beck commented:
Restarting PR build after weird failure, let's see whether it's transient.

szpak commented:
I enhanced comparison in FixedSet. Would you like anything else to change to make it accepted and merged?

daniel-beck commented:
@szpak Thanks, LGTM. @oleg-nenashev are you blocking?

oleg-nenashev reviewed (approved):
Weak 👍 for merging in the current state, will follow-up on the Default Case sensitivity option in a follow-up PR.

oleg-nenashev commented:
Thanks @szpak !
